src/routes/object/upsertObject.ts

```diff
@@ -0,0 +1,209 @@
+import { AuthenticatedRequest, Obj, ObjectMetadata } from '../../types/types'
+import { getOwner, getPostgrestClient, isValidKey, transformPostgrestError } from '../../utils'
+import { getConfig } from '../../utils/config'
+import { createDefaultSchema, createResponse } from '../../utils/generic-routes'
+import { deleteObject, headObject, initClient, uploadObject } from '../../utils/s3'
+
+const { region, projectRef, globalS3Bucket, globalS3Endpoint, serviceKey } = getConfig()
+const client = initClient(region, globalS3Endpoint)
+
+const upsertObjectParamsSchema = {
+  type: 'object',
+  properties: {
+    bucketName: { type: 'string', example: 'avatars' },
+    '*': { type: 'string', example: 'folder/cat.png' },
+  },
+  required: ['bucketName', '*'],
+} as const
+const successResponseSchema = {
+  type: 'object',
+  properties: {
+    Key: { type: 'string', example: 'projectref/avatars/folder/cat.png' },
+  },
+  required: ['Key'],
+}
+interface upsertObjectRequestInterface extends AuthenticatedRequest {
+  Params: FromSchema<typeof upsertObjectParamsSchema>
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export default async function routes(fastify: FastifyInstance) {
+  const summary = "Upsert the object, creates if it doesn't exist, otherwise updates"
+
+  const schema = createDefaultSchema(successResponseSchema, {
+    params: upsertObjectParamsSchema,
+    summary,
+    tags: ['object'],
+  })
+  fastify.post<upsertObjectRequestInterface>(
+    '/upsert/:bucketName/*',
+    {
+      schema,
+    },
+    async (request, response) => {
+      // check if the user is able to update the row
+      const authHeader = request.headers.authorization
+      const jwt = authHeader.substring('Bearer '.length)
+      const data = await request.file()
+      /* @ts-expect-error: https://github.com/aws/aws-sdk-js-v3/issues/2085 */
+      const cacheTime = data.fields.cacheControl?.value
+      const cacheControl: string = cacheTime ? `max-age=${cacheTime}` : 'no-cache'
+      let metadata: ObjectMetadata = {
+        mimetype: data.mimetype,
+      }
+      if (cacheTime) {
+        metadata.cacheControl = `max-age=${cacheTime}`
+      }
+
+      const { bucketName } = request.params
+      const objectName = request.params['*']
+
+      if (!isValidKey(objectName) || !isValidKey(bucketName)) {
+        return response
+          .status(400)
+          .send(createResponse('The key contains invalid characters', '400', 'Invalid key'))
+      }
+
+      const postgrest = getPostgrestClient(jwt)
+      const superUserPostgrest = getPostgrestClient(serviceKey)
+
+      let owner
+      try {
+        owner = await getOwner(jwt)
+      } catch (err) {
+        request.log.error(err)
+        return response.status(400).send({
+          statusCode: '400',
+          error: err.message,
+          message: err.message,
+        })
+      }
+
+      const objectResponse = await postgrest
```

Comment:
Can you use the [upsert function](https://supabase.io/docs/reference/javascript/upsert) of postgrest? Saves us a API call